File: backend/app/services/email_service.py
import random
import string
from datetime import datetime, timedelta
from models import db, PasswordVerify, EmailVerify, User
from sqlalchemy import desc
import requests
from config import Config

IRIS_DS_SERVER_URL = Config.IRIS_DS_SERVER_URL

def generate_reset_token(length=30):
    return ''.join(random.choices(string.ascii_letters + string.digits, k=length))

def send_authentication_email(username):
    user = User.query.filter_by(username=username).first()
    if not user:
        return {"message": "用戶不存在"}
    if not user.email:
        return {"message": "用戶未綁定 Email，請先綁定"}

    password_verify = PasswordVerify.query.filter_by(user_id=user.id).order_by(desc(PasswordVerify.valid_until)).first()
    current_time = datetime.utcnow()
    if password_verify and current_time <= password_verify.valid_until:
        return {"message": "驗證碼已發送，請前往電子信箱驗收，勿重複點取"}
    else:
        new_password_verify = PasswordVerify(
            password_verify_code=generate_reset_token(30),
            valid_until=datetime.utcnow() + timedelta(minutes=15),
            user_id=user.id
        )
        db.session.add(new_password_verify)
        db.session.commit()
        return {"message": "驗證信已發送，請重新設置"}

def send_email_verification(username):
    user = User.query.filter_by(username=username).first()
    if not user:
        return {"error": "用戶不存在"}
    if not user.email:
        return {"error": "用戶未綁定 Email，若需綁定請洽系統服務"}

    email_verify = EmailVerify.query.filter_by(user_id=user.id).order_by(desc(EmailVerify.valid_until)).first()
    current_time = datetime.utcnow()

    if email_verify and current_time <= email_verify.valid_until:
        return {"message": "驗證碼已發送，請勿重複點取"}
    else:
        new_email_verify = EmailVerify(
            email_verify_code=generate_reset_token(6),
            valid_until=datetime.utcnow() + timedelta(minutes=15),
            user_id=user.id
        )
        db.session.add(new_email_verify)
        db.session.commit()

    subject = "帳戶綁定確認"
    body_str = f'{new_email_verify.email_verify_code}'
    email_response = trigger_email(f"{IRIS_DS_SERVER_URL}/send-mail", user.email, subject, body_str)

    if "error" in email_response:
        return {"error": "Email 發送失敗"}
    return {"message": "驗證碼已發送，請檢查電子郵件"}

def trigger_email(url, recipient, subject, body_str):
    data = {
        "recipient": recipient,
        "subject": subject,
        "body": body_str
    }
    try:
        response = requests.post(url, json=data)
        if response.status_code == 200:
            return response.json()
        else:
            return {"error": f"Failed to send email, status code: {response.status_code}"}
    except Exception as e:
        return {"error": str(e)}

def send_email_code(username, code):
    user = User.query.filter_by(username=username).first()
    email_verify = EmailVerify.query.filter_by(user_id=user.id).order_by(desc(EmailVerify.valid_until)).first()
    if code != email_verify.email_verify_code:
        return {"error": "驗證碼錯誤"}
    if datetime.utcnow() > email_verify.valid_until:
        return {"error": "驗證碼已過期"}
    return {"message": "驗證成功，Email 已驗證"}

File: backend/nest-app/src/services/email.service.ts
import { Injectable, NotFoundException, BadRequestException } from '@nestjs/common';
import { InjectModel } from '@nestjs/sequelize';
import { User } from '../models/user.models';
import { EmailVerify } from '../models/email-verify.models';
import { PasswordVerify } from '../models/password-verify.models';
import { ConfigService } from '@nestjs/config';
import axios from 'axios';
import { Logger } from '@nestjs/common';

@Injectable()
export class EmailService {
  private readonly logger = new Logger(EmailService.name);
  constructor(
    @InjectModel(User)
    private userRepo: typeof User,
    @InjectModel(EmailVerify)
    private emailVerifyRepo: typeof EmailVerify,
    @InjectModel(PasswordVerify)
    private passwordVerifyRepo: typeof PasswordVerify,
    private configService: ConfigService,
  ) {}

  private generateCode(length: number): string {
    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';
    for (let i = 0; i < length; i++) {
      result += characters.charAt(Math.floor(Math.random() * characters.length));
    }
    return result;
  }

  async triggerEmail(to: string, subject: string, body: string) {
    const url = `${this.configService.get('IRIS_DS_SERVER_URL')}/send-mail`;
    try {
      const response = await axios.post(url, {
        recipient: to,
        subject,
        body,
      });
      return response.data;
    } catch (error) {
      console.error('Email 發送失敗:', error.response || error.message);
      throw new BadRequestException('Email 發送失敗');
    }
  }

  async sendEmailVerification(username: string) {
    const user = await this.userRepo.findOne({ where: { username } });
    if (!user) throw new NotFoundException('用戶不存在');
    if (!user.email) throw new BadRequestException('未綁定 Email');

    try {
      const existing = await this.emailVerifyRepo.findOne({
        where: { userId: user.id },
        order: [['validUntil', 'DESC']],
      });

      if (existing && new Date() <= existing.validUntil) {
        return { message: '驗證碼已發送，請勿重複點取' };
      }

      const code = this.generateCode(6);
      await this.emailVerifyRepo.create({
        emailVerifyCode: code,
        validUntil: new Date(Date.now() + 15 * 60 * 1000),
        userId: user.id,
      });

      await this.triggerEmail(user.email, '帳戶綁定確認', code);
      return { message: '驗證碼已發送，請檢查電子郵件' };
    } catch (error) {
      console.error('發送電子郵件驗證碼時發生錯誤:', error.message);
      throw new BadRequestException('發送電子郵件驗證碼時發生錯誤');
    }
  }

  async sendAuthenticationEmail(username: string) {
    const user = await this.userRepo.findOne({ where: { username } });
    if (!user) throw new NotFoundException('用戶不存在');
    if (!user.email) throw new BadRequestException('用戶未綁定 Email');

    const existing = await this.passwordVerifyRepo.findOne({
      where: { userId: user.id },
      order: [['validUntil', 'DESC']],
    });

    if (existing && new Date() <= existing.validUntil) {
      return { message: '驗證碼已發送，請勿重複點取' };
    }

    const code = this.generateCode(30);
    await this.passwordVerifyRepo.create({
      passwordVerifyCode: code,
      validUntil: new Date(Date.now() + 15 * 60 * 1000),
      userId: user.id,
    } as PasswordVerify);

    const link = `${this.configService.get('IRIS_DS_SERVER_URL')}/reset-password/${code}`;
    await this.triggerEmail(user.email, '帳戶綁定確認', link);
    return { message: '驗證信已發送，請重新設置' };
  }

  async sendEmailCode(username: string, code: string) {
    const user = await this.userRepo.findOne({ where: { username } });
    if (!user) throw new NotFoundException('用戶不存在');

    const record = await this.emailVerifyRepo.findOne({
      where: { userId: user.id },
      order: [['validUntil', 'DESC']],
    });

    if (!record || record.emailVerifyCode !== code) {
      throw new BadRequestException('驗證碼錯誤');
    }
    if (new Date() > record.validUntil) {
      throw new BadRequestException('驗證碼已過期');
    }

    return { message: '驗證成功，Email 已驗證' };
  }

  async sendRebindRequestEmail(username: string) {
    const user = await this.userRepo.findOne({ where: { username } });
    if (!user) throw new NotFoundException('找不到使用者');

    const body = `
有使用者申請重新綁定 Email。

使用者帳號：${user.username}
原本 Email：${user.email}

請客服人員儘速手動處理此請求。
    `;
    await this.triggerEmail('irisyen115@gmail.com', '【重新綁定 Email 請求】', body);
    return { message: '申請已送出，客服將會協助您重新綁定 Email' };
  }
}
